src/erlfmt.erl

```diff
@@ -214,24 +216,50 @@ replace_pragma_comment_block(_Prefix, [("%" ++ _) = Head | Tail]) ->
 replace_pragma_comment_block(Prefix, [Head | Tail]) ->
     [Head | replace_pragma_comment_block(Prefix, Tail)].
 
+% Format the minimum number of top-level forms
+% that cover the passed range.
+% Rationale: top-level forms is the smallest
+%            granularity we support now.
 -spec format_file_range(
     file:name_all(),
     erlfmt_scan:location(),
     erlfmt_scan:location(),
     [{print_width, pos_integer()}]
 ) ->
     {ok, string(), [error_info()]}
-    | {error, error_info()}
-    | {options, [{erlfmt_scan:location(), erlfmt_scan:location()}]}.
+    | {error, error_info()}.
 format_file_range(FileName, StartLocation, EndLocation, Options) ->
     {ok, Nodes, Warnings} = file_read_nodes(FileName, ignore),
-    format_range(FileName, StartLocation, EndLocation, Options, Nodes, Warnings).
+    format_enclosing_range(FileName, StartLocation, EndLocation, Options, Nodes, Warnings).
 
+-spec format_string_range(
+    string(),
+    erlfmt_scan:location(),
+    erlfmt_scan:location(),
+    [{print_width, pos_integer()}]
+) ->
+    {ok, string(), [error_info()]}
+    | {error, error_info()}.
 format_string_range(String, StartLocation, EndLocation, Options) ->
     FileName = "nofile",
     Pragma = proplists:get_value(pragma, Options, ignore),
     {ok, Nodes, Warnings} = read_nodes_string(FileName, String, Pragma),
-    format_range(FileName, StartLocation, EndLocation, Options, Nodes, Warnings).
+    format_enclosing_range(FileName, StartLocation, EndLocation, Options, Nodes, Warnings).
+
+format_enclosing_range(FileName, StartLocation, EndLocation, Options, Nodes, Warnings) ->
+    case format_range(FileName, StartLocation, EndLocation, Options, Nodes, Warnings) of
+        {options, PossibleRanges} ->
+            % Pick the largest range, so all intersected forms are coverd.
+            {Starts, Ends} = lists:unzip(PossibleRanges),
+            Start = lists:min(Starts),
+            End = lists:max(Ends),
+            Res = format_range(FileName, Start, End, Options, Nodes, Warnings),
+            ?assertNotMatch({options, _}, Res),
```

Comment:
Okay this looks great :D
Final nit: I think we shouldn't import the assert library into our non testing code. I think it is okay to do a manual case match here.

Reply:
Oh, ok. What is the rationale? Defensive programming is a thing too, and asserts give a more useful message (also when triggered the the intent is clearer than a mere error).

Reply:
My rationale is basically just to stick with current style of programming. I am sure there is a better rationale, but I haven't made the effort to look into it.  I hope that makes sense?